File: frontend/src/types/item.ts
import type { BaseQuery } from "./baseQuery";
import type { Category } from "./category";

export type ItemStatus = "available" | "unavailable";

export interface Item {
  id: number;
  name: string;
  description: string;
  category: Category;
  quantity: number;
  image: string;
  status: ItemStatus;
  createdAt: string;
  updatedAt: string;
}

export interface ItemCard {
  availableItems: number;
  unavailableItems: number;
}

export interface ItemRequest {
  name: string;
  description: string;
  categoryId: number;
  quantity: number;
  image?: File | null;
}

export interface ItemQuery extends BaseQuery {
  search?: string;
  categoryId?: number;
  status?: ItemStatus;
  unpage?: boolean;
}

export interface ItemError {
  name: string;
  description: string;
  categoryId: number;
  quantity: number;
}

File: backend/internal/model/item.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type ItemStatus string

const (
	ItemStatusAvailable   ItemStatus = "available"
	ItemStatusUnavailable ItemStatus = "unavailable"
)

type Item struct {
	ID          uint       `gorm:"primaryKey"`
	Name        string     `gorm:"type:varchar(255);not null;uniqueIndex:idx_name_deleted_at;index"`
	Description string     `gorm:"not null;type:text"`
	CategoryID  uint       `gorm:"not null;index"`
	Category    Category   `gorm:"constraint:OnDelete:CASCADE"`
	Quantity    int        `gorm:"not null"`
	Image       *string    `gorm:"type:varchar(255)"`
	Status      ItemStatus `gorm:"not null;type:enum('available', 'unavailable');default:unavailable"`
	CreatedAt   time.Time
	UpdatedAt   time.Time
	DeletedAt   gorm.DeletedAt `gorm:"index;uniqueIndex:idx_name_deleted_at"`
}
